docs: TSS-Library Proposal (#13808)
hashgraph/hedera-services: platform-sdk/docs/proposals/TSS-Library/TSS-Library.md

and confidentially yet verifiably distribute shares to multiple receivers.
It includes a distributed resharing protocol that preserves the public key but creates a fresh secret sharing of the secret key and hands it to a set of receivers,
which may or may not overlap with the original set of shareholders.

#### Overview
Here, the word participant describes parties producing the aggregate signature and aggregate public key.

Each scheme participant will receive its persistent EC private key from an external source and all participants' EC public keys.

This proposal covers the implementation of a tool similar to ssh-keygen to generate those keys, but the generation, persistence, distribution
and loading of those keys is outside the scope of this proposal.

Participants can hold one or more shares, each of which can be used to sign a message.
The goal is to generate an aggregate signature which is valid if a threshold number of individual signatures are combined.

Each participant brings their own Elliptic Curve (EC) key pair (private and public). They share their public keys with all other participants while securing their private keys.

> Before the protocol begins, all participants agree on the cryptographic parameters.

Author: Type of curve and group of the pairing to use for public keys

#### Implementation details
Before starting, all participants should agree on a `SignatureSchema` they will use.
The `SignatureSchema` defines the type of Curve and which Group of the Pairing is used for PublicKey Generation.

##### Input
* Participant's persistent EC Private key (Private to each participant)
* Number of participants (Public)
* Number of shares per participant (Public)
* A threshold value
* All participants' persistent EC public keys (Public)
* A predefined `SignatureSchema` (Public / Constant for all the network)

```
`Share`: An abstract concept having a unique identifier and an owner
|-  `PrivateShare`: Represents a share owned by the executor of the scheme. Contains a secret value (EC Private key) used for signing.
|-  `PublicShare`: Represents a share in the system. It contains public information that can be used to validate each signature and, when combined, to validate aggregate signatures.   
```

#####  Bootstrap Stage
Given

A participants directory:
```
e.g:
P   # shares
-----------------------------
P₁  5        P₁_EC_PublicKey
P₂  2        P₂_EC_PublicKey
P₃  1        P₃_EC_PublicKey
P₄  2        P₄_EC_PublicKey

```

A threshold value:
`e.g: t = 5`

> First, a `shareId`: `sid` is generated for each share. It's a unique identifier for each existent share. It is deterministic because the function we use to generate it is deterministic.

Author: thanks, added the clarification

It is necessary to assure they are values that: a) are unique per share, b) non-0, and c) can be used as input for the polynomial (They are from the same field of the selected curve)

And an ownership map: `ShareId`->`Participant`:

```
e.g:
sid₁	sid₂	sid₃	sid₄	sid₅	sid₆	sid₇	sid₈	sid₉	sid₁₀
P₁  	P₁  	P₁  	P₁  	P₁  	P₂  	P₂  	P₃  	P₄  	P₄  
```

##### 1. Create TssMessage
`TssMessage`: A data structure for distributing encrypted shares of a secret among all participants in a way that only the intended participant can see its part of the share.

> It includes validation information like a commitment to a secret share polynomial and a zk-SNARKs proof.

Author: Fixed! Thansk

The polynomial `Xₖ` is a polynomial with degree `t-1` (t=threshold) with the form:
`Xₖ = k + a₁x + ...aₜ₋₁xᵗ⁻¹`[ having: `a₁...aₜ₋₁`: random coefficients from `SignatureScheme.publicKeyGroup` and `k`'s EC field element. x is a field element, thus allowing the polynomial to be evaluated for each share id]

Each `xₛ = Xₖ(sidₛ)` constitutes a point on the polynomial.

Once the `xₛ` value has been calculated for each `ShareId`: `sidₛ`, the value: `Mₛ` will be produced by encrypting the `xₛ` using the `sidₛ` owner's public key.
The TssMessage will contain all the encrypted values for all shares.

![img.png](img.png)

A TssMessage class diagram:

![img_3.png](img_3.png)

##### Outside of scope
Using an established channel, each participant will broadcast a single message to be received by all participants
while waiting to receive other participants' messages. This functionality is critical for the protocol to work but needs to be handled outside the library.
Each participant will validate the received message against the commitment and the zk-SNARKs proof. Invalid messages needs to be discarded.

##### 2. Validation of TssMessage

> Each message can be validated against the commitment and the zk-SNARKs proof.

Author: Changed for NIZKs proofs, the type being used

  The signatures are compared against previous known signatures.
* Signature Verification: For each curve supported, 15 pseudorandom messages, a public key, Q, and a signature component (r, s) are supplied if some of the provided values are modified so that signature verification should fail.

Some ideas:
* [Cryptographic-Algorithm-Validation-Program](https://csrc.nist.gov/CSRC/media//Projects/Cryptographic-Algorithm-Validation-Program/documents/dss/ECDSAVS.pdf):
* https://csrc.nist.gov/CSRC/media/Events/lightweight-cryptography-workshop-2019/documents/papers/systematic-testing-of-lightweight-crypto-lwc2019.pdf

The proposal is to move forward with the implementation while creating a work group to discuss and collect ideas for testing and validating these features.
**Questions**:
* What other properties are not mentioned here?

### Unit Tests
One of the components that can be unit-tested is native support. SPI loading and failing. JNI wrapping.
Limitations: native-support-library tests must be executed in different environments to provide comprehensive code validation.

> ### Performance Testing

Reviewer: We should know that variance (min, max, average, median) time it takes to do the following: 

1. Genesis for a 4, 10, 20, 40, and 100 node network. 
2. Rekeying for a 4, 10, 20, 40, and 100 node network. 
3. Shares Amounts distributed randomly:  200, 500, 1000, 2000, 5000

we need to test each configuration 100 times to get confident mins, maxes, and a good average and median

Author: Added the notes, thanks

In multilanguage projects, Rust code will be compiled first using `Cargo` into a binary library and then packaged in the same jar.
While developing and on a developer’s machine, Rust code will be built for the architecture executing the build and packaged in the library’s jar, accessing the JNI wrapping code.
Once the code is merged to develop, the CI/CD pipeline compiles the Rust code into a binary library for each of the multiple supported platforms, packages it into a jar, and publishes it to Maven.
The dependency from the Maven repo will contain the Java code and the binary library cross-compiled for all supported architectures.

We will provide a library `swirlds-native-support` that will help load and use native code through JNI.
The low-level details of the build logic for this to work are outside the scope of this proposal.
A high overview is mentioned for the benefit of readers and will be provided by the DevOps Team alongside the RE Team.

### Libraries Specifications
#### Swirlds Crypto Pairings API
##### Overview
This API will expose general arithmetic operations to work with Bilinear Pairings and EC curves that implementations must provide.

##### Public API

> ###### `PairingsLibrabry`

Reviewer 2: ```suggestion
###### `PairingsLibrary`
```

  all points satisfy an equation, and all operations are performed modulo `p`. Some elliptic curves are pairing-friendly.
- **Bilinear Pairings**: These are mathematical functions used in cryptography to map two elements of different groups (in EC, the group is an elliptic curve) to a single value in another group
  in a way that preserves specific algebraic properties.
- **Fields**: Mathematical structures where addition, subtraction, multiplication, and division are defined and behave as expected (excluding division by zero).
- **Groups**: Sets equipped with an operation (like addition or multiplication) that satisfies certain conditions (closure, associativity, identity element, and inverses).
- **Share**: Represents a piece of the necessary public/private elements to create signatures. In TSS,
  a threshold number of shares is needed to produce an aggregate signature that the ledger public key can later verify.
- **Polynomial Commitment**: It enables verification of evaluations of the polynomial at specific points without revealing the entire polynomial.

> - **Participant**: Any party involved in the distributed key generation protocol.

Author: Done, skipped because we already had the definition of Shamir's secret sharing, but it is true that it is mentioned and not linked anywhere.

Additionally, participants will need access to each other's public key. While the generation of the public/private keys is included in this proposal,
the distribution aspect, the loading, and the in-memory interpretation from each participant are outside the scope of this proposal.

### Glossary
- **TSS (Threshold Signature Scheme)**: A cryptographic signing scheme in which a minimum number of parties (reconstruction threshold) must collaborate
  to produce an aggregate signature that can be used to sign messages and an aggregate public key that can be used to verify that signature.
- **Groth 21**: Publicly verifiable secret sharing and resharing schemes that enable secure and efficient distribution and management of secret shares,
  with many possible use cases supporting applications in distributed key generation and threshold signatures.
  Uses Shamir's secret sharing, ElGamal, and Zero-Knowledge Proofs.
- **Distribute key generation**: Aims to solve the problem of getting n parties able to cryptographically sign and verify signatures in the presence of some corruption threshold in a decentralized network. 

> To do so, this algorithm generates a public key, and a secret key of which no single party knows, but has some share of.

Reviewer 3: Can we change this to be something like:

"To do so, this algorithm generates a public key and a private key. The private key is distributed as key-splits such that no single party knows the private key."

The goal is to generate an aggregate signature which is valid if a threshold number of individual signatures are combined.

Each participant brings their own Elliptic Curve (EC) key pair (private and public). They share their public keys with all other participants while securing their private keys.
Before the protocol begins, all participants agree on the cryptographic parameters.
A participant directory is needed when initiating the protocol for the first time.
This directory includes the number of participants, each participant’s EC public key, and the shares they own.

Each participant generates a random private key and distributes it among the others.
This allows for verification and aggregation of signatures produced by all shares.
A value derived from this random private key is created for each share using mathematical functions. Each value is encrypted with the share owner's public key, ensuring only the intended recipient can read it.
When a message is created, all encrypted values are included. Only the intended receivers can decrypt their respective portions of the secret element.
This setup allows participants to share secret information securely.

> The message also contains additional information necessary for signature validation.

Author: Each participant can act as a dealer or receiver.

When generating the keys, they act as dealers, creating the shares for the other participants out of the secret shares they own. They cannot "spy" on other participant's secrets as they don't have access to them. If they are malicious, they could send an invalid message, which other non-malicious participants would detect through the proof and commitments.
When acting as receivers, given that each share is encrypted under each participant's public key, they need access to other participants' private keys to decrypt and "spy" on the shares belonging to other participants.
We can mitigate the risk with a key rotation policy.

@rsinha, please jump in if I need to correct something in the answer.

> The message also contains additional information necessary for signature validation.

Reviewer 4: That's right. As long as the adversary has knowledge of fewer than a threshold number of decryption keys, they cannot recover enough information to start forging threshold signatures -- they may learn the shares of the nodes whose keys they have compromised, but that's not enough to recover the full secret.

> The message also contains additional information necessary for signature validation.

Reviewer 2: Thank you for that explanation.  I'm still missing something (which suggests the document could use some clarifying text).

What's still not clear to me, is how the nodes are prevented from obtaining a threshold number of unencrypted shares.
Perhaps it's just a missing step or word confusion; I'll list what I think is happening, based on the proposal description, below.
1. Nodes gossip random private keys.
   - "Each participant generates a random private key and distributes it among the others."
1. A dealer (one, many?) receives all the random values, generates a share for each random value
   - "A value derived from this random private key is created for each share using mathematical functions."
1. The dealer, after deriving the value, encrypts the share
   - "Each value is encrypted with the share owner's public key, ensuring only the intended recipient can read it."
      - Couldn't a malicious actor inspect all of the values here, and possibly encrypt chosen share values?
      - Secondary, is every node doing this, in concert, or a single dealer?
         - I am assuming every node completes this step.
1. Each node, having calculated _all_ of the encrypted shares, gossips out the encrypted shares.
   - "When a message is created, all encrypted values are included."

This next bit is where it gets confusing.
 - "Upon receiving a threshold number of messages, each participant decrypts the information encrypted with their public key, aggregates it, and generates a private key for each owned share."

Would that mean each node is computing a _partial_ share for each other node?  That might be the part I am missing, but I'm not certain.

> The message also contains additional information necessary for signature validation.

Author: Let me work on the rewording. I see the confusion.

> The message also contains additional information necessary for signature validation.

Author: is this more clear:
```
Each participant generates portions of a secret share and distributes them among the other participants using the following process:

1. A random private key is created and mathematically split into a known number of total shares.
2. Each portion is encrypted with the share owner's public key, ensuring only the intended recipient can read it.
3. A message is created that includes all encrypted values so that only the intended recipients can decrypt their respective portions of the secret share.

This setup allows participants to share secret information securely. The message also contains additional information necessary for its validation.

Upon receiving a threshold number of messages, each participant:

1. Validates the message and decrypts the information encrypted with their public key.
2. Aggregates the decrypted information to generate a private key for each owned share.
3. Retrieves a public key for each share in the system to validate signatures.

Individual signing can then begin. Participants use the private information of their shares to sign messages.

An aggregate signature is created when signatures from at least the threshold number of parties are combined. This aggregate signature can be validated using the combined value of the public shares in the directory.

The process restarts whenever the number of participants or the shares assigned to each change. However, the initially generated group public key remains unchanged to maintain consistency. New secret information for shares is created using existing data, ensuring the aggregate signature can still be verified with the original group public key.
```

> The message also contains additional information necessary for signature validation.

Reviewer 2: Much clearer.  We might add a sub-point to express how these independent actions are effectively coordinated.  Perhaps something like the following (I am somewhat guessing that this is the case; please correct if I have it wrong):
> 1. A random private key is created and mathematically split into a known number of total shares.
>>    * These shares are related to shares from other participants through a pre-arranged polynomial commitment.

##### Input
* Participant's persistent EC Private key (Private to each participant)
* Number of participants (Public)
* Number of shares per participant (Public)
* A threshold value
* All participants' persistent EC public keys (Public)
* A predefined `SignatureSchema` (Public / Constant for all the network)

```
`Share`: An abstract concept having a unique identifier and an owner
|-  `PrivateShare`: Represents a share owned by the executor of the scheme. Contains a secret value (EC Private key) used for signing.
|-  `PublicShare`: Represents a share in the system. It contains public information that can be used to validate each signatures.   
```

#####  Bootstrap Stage
Given

A participants directory:
```
e.g:
P   # shares
-----------------------------
P₁  5        P₁_EC_PublicKey
P₂  2        P₂_EC_PublicKey
P₃  1        P₃_EC_PublicKey
P₄  2        P₄_EC_PublicKey

```

A threshold value:
`e.g: t = 5`

First, a `shareId`: `sid` is generated for each share. It's a unique, contiguous starting from 1 identifier for each existent share. It is deterministic because the function we use to generate it is deterministic.

> It is necessary to assure they are values that: a) are unique per share, b) non-0, and c) can be used as input for the polynomial (They are from the same field of the selected curve)

Author: The share IDs are part of the proof and commitment. All nodes have to use the same values for the shared IDs. A malicious node using different values will be detected when validating the tssMessage against the proof and the commitment.
But if there exists a specific set of values such as they would affect the validation, making it take longer than it should, I will defer the answer to @rsinha.

> It is necessary to assure they are values that: a) are unique per share, b) non-0, and c) can be used as input for the polynomial (They are from the same field of the selected curve)

Reviewer 4: As for security, adversarial nodes are free to choose low or zero entropy values for the protocol. There will still be sufficient entropy in the entire protocol, from the honest nodes. As for correctness, the choice of the value does not impact whether any downstream checks fail, as long as the protocol is followed -- if a node deviates from the protocol and produces invalid messages, those are certainly detected. Happy to expand further.

> It is necessary to assure they are values that: a) are unique per share, b) non-0, and c) can be used as input for the polynomial (They are from the same field of the selected curve)

Reviewer 2: Thank you for the explanation, that clarifies somewhat.

It is not certain (there may be "show-the-way" concerns), but it might be helpful to include a section on how this proposed approach handles malicious actors; and include answers like this in that section.
Just a thought.

> It is necessary to assure they are values that: a) are unique per share, b) non-0, and c) can be used as input for the polynomial (They are from the same field of the selected curve)

Author: Done

```
`k`  -  A random EC Private Key for the participant
`n`  -  The number of total shares across all participants
`sidᵢ`- The shareId of the share i
`Xₖ` -  A polynomial with certain properties given a specific secret k
`sᵢ` -  A point in the polynomial Xₖ for sidᵢ
```

Then, each shareholder will produce `n` (n=total number of shares) values `Xₛ` by evaluating a polynomial Xₖ at each `ShareId`: `sidᵢ` in the ownership map.

The polynomial `Xₖ` is a polynomial with degree `t-1` (t=threshold) with the form:
`Xₖ = k + a₁x + ...aₜ₋₁xᵗ⁻¹`[ having: `a₁...aₜ₋₁`: random coefficients from `SignatureScheme.publicKeyGroup` and `k`'s EC field element. x is a field element, thus allowing the polynomial to be evaluated for each share id]

Each `sᵢ = Xₖ(sidᵢ)` constitutes a point on the polynomial.

Once the `sᵢ` value has been calculated for each `ShareId`: `sidᵢ`, the value: `Cᵢ` will be produced by encrypting the `sᵢ` using the `sidᵢ` owner's public key.

![img.svg](img.svg)

The TssMessage will contain all the encrypted values for all shares.

###### Generation of the Polynomial Commitment
Secret sharing poses a problem for the receiver: did it get a correct share? 

> The dealer may give her a bad share that does not correspond to the dealing, or give so many fake shares to different receiver that they do not correspond to a real dealing. 

Reviewer 2: Minor wording fix:
```suggestion
The dealer may send a bad share that does not correspond to the dealing, or send so many fake shares to different receivers that the aggregate does not correspond to a real dealing. 
```

We include a Feldman commitment to the polynomial.

```
 g = a point generator of `SignatureScheme.publicKeyGroup`
 aₒ...aₜ₋₁ = coefficients of the polynomial being commited to.
```

> For each coefficient in the polynomial `Xₖ` `a₍ₒ₎` to `a₍ₜ₋₁₎`, computes a commitment value by calculating: `gᵢ * aᵢ ` (g multiplied by polynomial coefficient `a₍ᵢ₎` )

Reviewer 2: ```suggestion
For each coefficient in the polynomial `Xₖ` `a₍ₒ₎` to `a₍ₜ₋₁₎`, compute a commitment value by calculating: `gᵢ * aᵢ ` (g multiplied by polynomial coefficient `a₍ᵢ₎` )
```

###### Generation of the NIZKs proofs

Generate a NIZK proof that these commitments and the encrypted shares correspond to a valid sharing of the secret according to the polynomial.

Given the input:

* witness: private data for the proof, including the actual share and the randomness used in the encryption process and private shares(sᵢ).

> * statement: The information to proof (sidᵢs,public_keys, polynomial_commitment, encrypted_shares )

Reviewer 2: ```suggestion
* statement: The information to prove (sidᵢs,public_keys, polynomial_commitment, encrypted_shares )
```

* rng: source of randomness

1. Serialize statement
2. Initialize a pseudorandom number generator using the `SHA256` hash of (1)
3. `x` = Random out of the random generator in (2)  

> 4. Generate random values `α`, `ρ` from the `SignatureScheme.field` out of rng

Reviewer 2: Is there an ordering requirement here?

Author: In the generation of `α`  and `ρ` values?
I understand there is not.
@rsinha ?
I don't see anything in the code that makes that true. Do you agree?

Reviewer 2: To be specific, I'm asking if all nodes need to use the _same order_, to maintain hash consistency.

Reviewer 4: The statement being serialized and hashed only includes elements computed by the dealer. While there is an ordering amongst those elements as defined by the serialization of that statement, there is no ordering "across" nodes, as the statement is produced by a single node.

5. `g` = `SignatureScheme.publicKeyGroup.generator`
  5.1    `F = g^ρ`
  5.2    `A = g^α`

> 6. Calculate `Y` by transforming element of the statements (public keys, sids) multiplied by `ρ`, then adding `A`

Reviewer 2: ```suggestion
6. Calculate `Y` by transforming elements of the statement (public keys, sids) multiplied by `ρ`, then adding `A`
```

    21.1.  `term` = `public_key` multiplied by (`proof.z_r` * (`x` raised to the power of `sid`))

    21.2.  `inner` =`inner` + `term`
22. `rhs` = `inner` + (`g` multiplied by `proof.z_a`)
23. If `lhs` != `rhs`: Return False
24. Return True

##### Generating Participant's Private Shares & Ledger Id
Given Participant's persistent EC PrivateKey and precisely `t` number of validated messages (t=threshold)
each participant will decrypt all `Cᵢ` to generate an aggregated value `sᵢ` that will become a  `SecretShare(sidᵢ, sᵢ)` for each `ShareId`: `sidᵢ` owned by the participant.

**Note:** All participants must choose the same set of valid `TssMessages` and have a threshold number of valid messages.

![img_1.svg](img_1.svg)

Also, we will extract a `PublicShare` for each `ShareId`: `sidᵢ` in the directory from the list of valid messages.
The PublicShare for share `s` is computed by evaluating each polynomial commitment in the common set of messages at `sidᵢ` and then aggregating the results.

At this point, the participant executing the scheme can start signing, sharing signatures, and validating individual signatures produced by other parties in the scheme.

#####  Rekey Stage

> This Section describes the rekey process executed each time some of the scheme’s parameters change, such as the number of participants or the number of shares assigned to each participant.
> The rekeying process can happen, too, if the parameters are the same as the genesis procedure.

Reviewer 2: A bit awkward.
Suggested reword:
```suggestion
This Section describes the rekey process executed each time some of the scheme’s parameters change, such as the number of participants or the number of shares assigned to each participant.
This rekeying process may also be conducted for reasons other than parameter changes; to rotate shares, for example.
```

The rekeying process is similar to the bootstrap process, but it starts with the previous list of owned private `SecretShare`.
The main difference with the genesis stage is that every participant generates a `TssMessage` out of each previously owned `SecretShare`.

![img_4.svg](img_4.svg)

Once finished, the list of `SecretShare`s will be updated but the previously generated aggregate public key remains the same.

#####  Sign Stage
After genesis or rekeying stages, the library can sign any message.
Using each `SecretShare` owned by the participant, a message can be signed, producing a `TssSignature`

> Multiple signatures can be aggregated to create an aggregate `TssSignature`. An aggregate `TssSignature` can be validated against the LedgerId if
> `t` (t=threshold) valid signatures are aggregated. If the threshold is met and the signature is valid, the library will respond with true; if not, it will be false.

Reviewer 2: ```suggestion
Multiple signatures can be aggregated to create an aggregate `TssSignature`. An aggregate `TssSignature` can be validated against the LedgerId (public key) if
`t` (t=threshold) valid signatures are aggregated. If the threshold is met and the signature is valid, the library will respond with true; if not, it will respond with false.
```

#### Security Considerations
As long as an adversary Participant knows fewer than a threshold number of decryption keys, they cannot recover enough information to start forging threshold signatures.
Adversarial participants may learn the shares of the participants whose keys they have compromised, but more is needed to recover the secret.

> For security, adversarial parties are free to choose low or zero entropy values for protocol inputs such as shareIds.
> There will still be sufficient entropy in the entire protocol from the honest nodes.

Reviewer 2: ```suggestion
For security, adversarial parties might choose low or zero entropy values for protocol inputs such as shareIds, but here will still be sufficient entropy in the overall protocol from honest nodes.
```

#### Summary diagram
The following diagram exposes the steps necessary for bootstrapping and using the library. In orange are all the steps where the library is involved.
The rest of the steps need to happen but are outside the scope of the described library.

![img_2.png](img_2.svg)

### Architecture
To implement the functionality detailed in the previous section, the following code structure is proposed:

![img_5.png](img_5.svg)

1. **TSS Lib**: The consensus node will use the TSS library to create shares, create TSS messages to send to other nodes,
   assemble shared public keys (ledgerId), and sign the block node Merkle tree hash.
2. **Bilinear Pairings Signature Library**: This library provides cryptographic objects (PrivateKey, PublicKey, and Signature)
   and operations for the block node and consensus node to sign and verify signatures. The consensus node uses this library indirectly through the TSS Library.

> 3. **Bilinear Pairings API**: Layer of abstraction to be included at compilation time providing the cryptography primitives
>    and the arithmetic operations for working with a specific EC curve and the underlying Groups, Fields, and Pairings
>    and minimizes the impact of changing or including different curve implementations.
> 4. **Bilinear Pairings Impl**: This is an implementation of the Bilinear Pairings API that will be loaded at runtime using Java’s SPI.
>    Multiple implementations can be provided to support different types of curves. It allows changing or switching dependencies.

Reviewer 2: Some mild awkward wording.  Suggested reword:
```suggestion
3. **Bilinear Pairings API**: An API definition for the cryptography primitives
   and arithmetic operations required to work with a specific EC curve and the
   underlying Groups, Fields, and Pairings. This API minimizes the impact of
   changing to different curve implementations.
4. **Bilinear Pairings**: An implementation of the Bilinear Pairings API that will be loaded at runtime using Java’s service provider (SPI) mechanism.
   Multiple implementations can be provided to support different types of curves and
   the Java service provider approach facilitates easily changing between
   implementations and dependencies.
```

Author: Thanks!

5. **Native Support Lib**: Provides a set of generic functions loading native libraries in different system architectures when packaged in a jar
   using a predefined organization so they can be accessed with JNI.
6. **[Arkworks](https://github.com/arkworks-rs)**: A Rust ecosystem for cryptography. Our implementation uses it as the library responsible for elliptic curve cryptography.

> 7. **EC-Key Utils** is a utility module that enables the node operator to generate a bootstrapping public/private key pair.

Reviewer 2: ```suggestion
7. **EC-Key Utils** is a utility module that enables the node operator to generate a bootstrap public/private key pair.
```

### Module organization and repositories
![img_6.png](img_6.svg)
1. **hedera-cryptography**: This is a separate repository for hosting cryptography-related libraries.

>    It is necessary to facilitate our build process, which includes Rust libraries. It also provides independent release cycles between consensus node code and block node code.

Reviewer 2: ```suggestion
   It is necessary to facilitate our build process, which includes Rust libraries. It also provides independent release cycles between consensus node code and other library users.
```

2. **swirlds-native-support**: Gradle module that enables loading into memory compiled native libraries so they can be used with JNI.
3. **swirlds-crypto-tss**: Gradle module for the TSS Library.
   This library's only client is the consensus node, so it will be in the hedera-services repository, under the `platform-sdk` folder.
4. **swirlds-crypto-signatures**: Gradle module for the Bilinear Pairings Signature Library.
5. **swirlds-crypto-pairings-api**: Gradle module for the Bilinear Pairings API. Minimizes the impact of adding or removing implementations.
6. **swirlds-crypto-altbn128**: Gradle module that will implement the Bilinear Pairings API using alt-bn128 elliptic curve.
   That curve has been chosen due to EVM support. The arkworks rust library will provide the underlying cryptography implementation.

>    The module will include Java and Rust codes that will be compiled for all possible system architectures and distributed in a jar under a predefined structure.

Reviewer 2: Are these "Gradle" modules or "Java" modules?  I believe, based on structure, that these are "Java" modules (e.g. defined by a module-info.java file).
Gradle just has support for building the java modules; I think.

Author: They are gradle modules. @jjohannes will help create the necessary plug-ins to make this happen. I have already implemented a rustic version to determine if it was possible based on some previous work by @nathanklick .
Jendrik already found some possible utils plugins to help with this:
https://ysb33rorg.gitlab.io/rust-gradle-plugin/0.2/docs/product-documentation.html

>    The module will include Java and Rust codes that will be compiled for all possible system architectures and distributed in a jar under a predefined structure.

Reviewer 2: I do not think we will build for all _possible_ architectures (are we supporting obsolete things like System/360 and DEC Alpha, or new architectures like RISC-V?).  Might be better to describe _supported_ architectures.
```suggestion
   The module will include Java and Rust code that will be compiled for all supported system architectures and distributed in a jar with a predefined structure.
```

### Handling of multilanguage modules and native code distribution

> The code provided by this solution will include multilanguage modules (rust + java). Rust code needs to be compiled into binary libraries and accessed through JNI.
> Given the non-GA state of Project Panama, this proposal does not consider it, but it has not been discarded for future change proposals.

Reviewer 2: Possible wording improvement:
```suggestion
The software provided by this approach will require multi-language modules (rust + java). Rust code must be compiled into binary libraries and accessed through JNI.
Given the immature state of development, Project Panama is not considered for this proposal; but Project Panama may be considered for future development.
```

There are two possible ways of loading libraries and accessing them through JNI

> 1) Libraries are installed in the SO and referenced through the classpath.

Reviewer 2: The `SO` term should probably be defined.
This is what I _think_ was intended (but I'm guessing a little here).
```suggestion
1) Libraries are installed on the system as shared object(SO) libraries and found via the classpath and system library path.
```

> 2) Distributed with the application jars, unzipped, and loaded in runtime.

Reviewer 2: ```suggestion
2) Distributed with the application jars, unpacked, and loaded at runtime.
```

We want to ensure that dependent software does not require the installation of any additional dependencies other than those distributed in the jars, so we are choosing option 2.

In multilanguage projects, Rust code will be compiled first using `Cargo` into a binary library and then packaged in the same jar.
While developing and on a developer’s machine, Rust code will be built for the architecture executing the build and packaged in the library’s jar, accessing the JNI wrapping code.
Once the code is merged to develop, the CI/CD pipeline compiles the Rust code into a binary library for each of the multiple supported platforms, packages it into a jar, and publishes it to Maven.

> The dependency from the Maven repo will contain the Java code and the binary library cross-compiled for all supported architectures.

Reviewer 2: I have a question about this item.

Would it make better sense (and be more secure) to only include one architecture per jar, and select the relevant jar for a given architecture?

We could then produce OCI images for each supported architecture containing only the native code for that architecture, and thus avoid distributing incompatible code that might have potential for issues in the system.

Author: This build process was defined with @nathanklick.
The benefit is that you maintain the "build once, run everywhere" property. (where everywhere in this case is just supported architectures, as you mentioned above)

The risk you mention is mitigated by the fact that the code will make the right decision regarding the library to load and only unpackage the right binary for the architecture in execution.

Of course, if somebody gains access to the node, unpackages the content of this jar, and executes the wrong file, bad things will happen. Is there any other scenario you think of where the distributed binary can create a problem if it remains packaged?

@hendrikebbers had a similar proposal as yours ragarding this topic.
I guess there are pros and cons to both strategies.

I think we should go with the one defined here, but if you still strongly oppose it, I can arrange a meeting with all interested parties.

Reviewer 2: It's not a strong opposition.  More of something to bring up with our product security experts. It may be possible, in rare circumstances, to make running code try to load a bad library (that's been done to cause failures in other systems, if I recall correctly).

Run-anywhere from a single jar is less valuable, in my opinion, with native code and OCI images (which are system specific regardless), and we do preserve that capability when using different dependencies by system (which we have regardless, because some of our own dependencies are organized as different jars for each architecture/system)

This suggestion would match a common pattern in open source, but you are correct that each approach has different benefits and costs.

**Link**: [Curve.java](pairings-api%2FCurve.java)

**Note**: This enum should list all possible types so changes in the implementation don't change the API, but as of now, the number is limited.

###### `BilinearPairing`

**Description**: This class provides access to each group (G₁, G₂) for a specific Pairing and the FiniteField associated with the curve.
It provides an initialization method to load all necessary dependencies for the library to work. Callers are requested to invoke that method before start using the library.

**Link**:  [BilinearPairing.java](pairings-api%2FBilinearPairing.java)

###### `Field`
**Description**: This is a factory interface responsible for creating FieldElement, which are scalars belonging to the field represented by this instance.

**Link**: [Field.java](pairings-api%2FField.java)

###### `FieldElement`

**Description**: An interface representing an element within a field, providing methods for basic arithmetic operations.

**Link**: [FieldElement.java](pairings-api%2FFieldElement.java)

###### `Group`

> **Description**: This class provides methods to obtain elements belonging to the group represented by the instance.

Reviewer 2: Should this be consistent with the other interface descriptions?
```suggestion
**Description**: An interface to define methods that obtain elements belonging to the group(s) supported by an implementation.
```

**Description**: A pairings signature scheme can be implemented with different types of curves and group assignment configurations.
For example, two different configurations might consist of a BLS_12_381 curve using G1 of the pairing to generate public key elements or G2 for the same purpose.
###### `PairingPrivateKey`
**Link**: [PairingPrivateKey.java](signature-lib%2FPairingPrivateKey.java)

**Description**: A private key generated using the pairings API

###### `PairingPublicKey`
**Link**: [PairingPublicKey.java](signature-lib%2FPairingPublicKey.java)

**Description**: A public key generated using the pairings API
###### `PairingSignature`
**Link**: [PairingSignature.java](signature-lib%2FPairingSignature.java)

**Description**: A signature generated with the private key that can be verified with the public key

###### `GroupAssignment`
**Link**: [GroupAssignment.java](signature-lib%2FGroupAssignment.java)

**Description**: An enum to clarify which group public keys and signatures are in, for a given SignatureSchema

##### Implementation Note
The serialization of the elements in this module adds a byte to represent the combination of Curve type and group assignment.

> ##### Examples

Reviewer 2: In the examples, rather than shorthand variable names, perhaps name each variable to describe what it is in an example.
For example:
instead of `s.verify(pbk, message)`  write `signatureToVerify.verify(pairingPublicKey, messageSigned)`.
That would be much clearer for the reader, trying to follow through multiple examples in series.

Author: Makes sense, thanks for pointing it out.

###### Generating a SignatureSchema
```java
SignatureSchema signatureSchema = SignatureSchema.from(Curve.ALT_BN128, GroupAssignament.G1_PUBLIC_KEYS);
```
###### Generating a Private Key
```java
   PairingPrivateKey pk = PairingPrivateKey.create(signatureSchema, new SecureRandom());
   
```
###### Generating a Public Key
```java
   PairingPublicKey pbk = PairingPublicKey.create(pk);
   
```
###### Generating a Signature
```java
    byte[] message = new byte[]{};
    PairingSignature s = PairingSignature.sign(pk, message );
   
```
###### Verifying a Signature
```java
static{
    s.verify(pbk, message);
}
```
##### Constraints
This module will not depend on hedera-services artifacts, so it cannot include logging, metrics, configuration, or any other helper module from that repo.
##### Dependencies
swirlds-cryptography-pairings-API and runtime implementation
##### Other considerations

> Some unknowns are worth investigating in a follow-up task:

Reviewer 2: Was there supposed to be a list following this line?

Author: Should have removed this list

#### Swirlds Crypto TSS Library
##### Overview
This library implements the Groth21 TSS-specific primitives.

##### Constraints
Once built, this library will depend on artifacts constructed in other repositories.

> This library would accept Integer.MAX_VALUE -1 participants.

Reviewer 2: ```suggestion
This library should accept Integer.MAX_VALUE -1 participants.
```

Author: `This library accepts Integer.MAX_VALUE -1 participants.`

I think this is an important statement given that it also means we can support Integer.MAX_VALUE -1 nodes

Reviewer 2: Your statement is reasonable, just perhaps a bit in excess of realistic need.

In the real world we'll not likely get anywhere near 2 billion nodes (are there even that many servers in existence right now?) for a very long time.

##### Public API:
###### `TssService`
**Description**: This class handles all tss specific operations.

**Link**: [TssService.java](tss%2FTssService.java)

###### `TssParticipantDirectory`
**Description**: This class holds all information about the participants in the scheme. Including: participants' EC public keys, public shares, private shares, number of shares. 

**Link**: [TssParticipantDirectory.java](tss%2FTssParticipantDirectory.java)

###### `TssMessage`
**Description**: This class is used in the exchange of secret information between participants. Contains an encrypted message, a polynomial commitment, and a cryptographic proof that can be used to validate this message.

**Link**:[TssMessage.java](tss%2FTssMessage.java)

###### `TssShareId`

> **Description**: This class represents the unique identification of a share in the system. It a FieldElement, so it can be used as input in the polynomial.

Reviewer 2: Might be missing a few words around `It a FieldElement` and `the polynomial`.
It's not clear to me what was actually intended for that sentence currently.

Author: Half worked through a previous suggestion. should be:
`It contains a FieldElement, so it can be used as input in the polynomial.`

**Link**:[TssShareId.java](tss%2FTssShareId.java)

###### `TssPrivateShare`
**Description**: A record that contains a share ID, and the corresponding private key.

**Link**:[TssPrivateShare.java](tss%2FTssPrivateShare.java)

###### `TssPublicShare`
**Description**: A record that contains a share ID, and the corresponding public key.

**Link**:[TssPublicShare.java](tss%2FTssPublicShare.java)

###### `TssShareSignature`
**Description**: Represents a signature created from a TSSPrivateShare.

**Link**:[TssShareSignature.java](tss%2FTssShareSignature.java)

##### Examples:
###### 1. Bootstrapping
```java
   TssService service = new TssService(signatureScheme, new Random());

>    PairingPrivateKey persistentParticipantKey = Requirements.loadECPrivateKey();

Reviewer 2: Is this always singular?  Could this also be a list, like the public keys on the next line?

Author: The private key is unique per participant (Node).
The public keys are all the public keys for all the participants.

e.g.
In a 3-node setup, node 1 will have its privateKey and a list of 3 publicKeys.

   List<PairingPublicKey> persistentPublicKeys = Requirements.loadECPublicKeys();
   TssParticipantDirectory participantDirectory =
           TssParticipantDirectory.createBuilder()

>            .self(/**/ 0, persistentParticipantKey)

Reviewer 2: Missing content within the comment markers?

```java
    Set<TssMessage> agreedValidMessages = /*Some previously agreed upon same set of valid messages for all participants*/
    //Get Private Shares    
    List<TssPrivateShare> privateShares = service.decryptPrivateShares(participantDirectory, agreedValidMessages);
    //Get Public Shares
    List<TssPrivateShare> publicShares = service.computePublicShare(participantDirectory, agreedValidMessages);

```
######  Sign
```java
   byte[] message = Requirements.messageToSign();
   List<TssShareSignature> signatures = service.sign(privateShares, message);

```

######  Aggregate Signatures

```java
static {
    List<TssShareSignature> signatures = Requirements.loadSignatures();

    //Validation of individual signatures
    List<TssShareSignature> validSignatures = new ArrayList<>(signatures.size());
    for (TssShareSignature s : signatures) {
        if (service.verify(participantDirectory, publicShares, s)) {
            validSignatures.add(s);
        }
    }

    //Producing an aggregate signature
    PairingSignature aggregateSignature = service.aggregate(participantDirectory, validSignatures);
}
```

#### Swirlds Native Support
##### Overview

> It helps other modules to load binary libraries packaged inside the library's jar into memory.

Reviewer 2: ```suggestion
This library provides classes that assist other modules to load native libraries packaged in dependency jars.
```

> It provides a way to describe the library for different architectures/ OS.

Reviewer 2: ```suggestion
This library provides a pattern to describe libraries for different architectures and operating systems.
```

##### Constraints
This module will not depend on hedera-services artifacts, so it cannot include logging, metrics, configuration, or any other helper module from that repo.

##### Public API
###### `OperatingSystem`
**Description**: An enum class listing all supported OS.

**Link**: [OperatingSystem.java](native%2FOperatingSystem.java)

###### `Architecture`
**Description**: An enum class listing all supported architectures.

**Link**: [Architecture.java](native%2FArchitecture.java)

###### `LibraryDescriptionEntry`
**Description**: Given that the compilation of binary libraries produces files with different names under different OS and architectures, we need a way to assemble a catalog for all possible forms our library will take.
A record of 3 elements that defines the name of the binary file of the library to load in a specific system architecture and OS.

###### `LibraryDescription`

> **Description**: A description of the library in all possible systems.

Reviewer 2: ```suggestion
**Description**: A description of the library in all supported systems.
```

###### `LibraryLoader`
**Description**: Helper class that will load a library for the correct system

##### Example

```java
static {
    List<LibraryDescriptionEntry> entries = Lists.of(
            new LibraryDescriptionEntry(OperatingSystem.DARWIN, Architecture.ARC64, "libhedera_bls_jni.dylib"),
            new LibraryDescriptionEntry(OperatingSystem.LINUX, Architecture.AMD64, "libhedera_bls_jni.so") /*,... Others*/);
    LibraryDescription libHederaDescription = new LibraryDescription(entries);
    try {
        LibraryLoader.load(libHederaDescription);
    } catch (IOException e){
        //Handle library loading problems
    }
}
```

##### Implementation details
Loading a library for a system architecture where no description is provided will result in a runtime exception.

Under the hood, LibraryLoader will make use of:

> https://github.com/hashgraph/full-stack-testing/blob/c3fd5602525145be132770116f5bb5a1a1922dea/fullstack-core/fullstack-base-api/src/main/java/com/hedera/fullstack/base/api/resource/ResourceLoader.java

Reviewer 2: This is expressly a test file, so perhaps we mean to say "a class similar to" this will be used?

Author: indeed.

```
resources/software
    ├── LIBRARY_VERSION (Contains the version number or information for the original native library)
    ├── darwin
    │   ├── amd64
    │   │   ├── native_lib.dylib **
    │   │   └── jni_bindings.dylib
    │   └── arm64
    │   │   ├── native_lib.dylib **
    │   │   └── jni_bindings.dylib
    ├── linux
    │   ├── amd64
    │   │   ├── native_lib.so **
    │   │   └── jni_bindings.so
    │   ├── arm
    │   │   ├── native_lib.so **
    │   │   └── jni_bindings.so
    │   └── arm64
    │   │   ├── native_lib.so **
    │   │   └── jni_bindings.so
    └── windows
        └── amd64
            ├── native_lib.dll **
            └── jni_bindings.dll

** NOTE: native_lib is only required if not statically linked to the native JNI binding library
```

## Test Plan
Since cryptographic code is often difficult to test due to its complexity and the lack of a test oracle,
we should design our test cases based on the cryptographic properties that these implementations should satisfy.

**Properties**:

Protection of private keys. Our scheme protects the secrecy of the group’s and members’ private keys; otherwise, private keys can generate only one group signature.

> Unforgibility of group signature. Our scheme ensures that (a) any t or more than t members can work

Reviewer 2: ```suggestion
Unforgeability of group signature. Our scheme ensures that only a subset containing t, or more than t, members can work
```

> together to generate a valid group signature, and (b)fewer than t members cannot generate a valid group signature.

Reviewer 2: ```suggestion
together to generate a valid group signature, and (b) fewer than t members cannot generate a valid group signature.
```

Fixed length of threshold signature. Our scheme ensures that the size of a threshold signature is fixed (i.e., not depending on the number of signers).

> Efficiency of verification. Verifying a group signature is based on the group’s public key.

Reviewer 2: ```suggestion
Efficiency of verification. Verifying a group signature is based on the group’s public key and may be computed efficiently in an EVM context.
```

> * Generation of Private and Public Key Pairs: Tests the generation of key pairs for correctness by having the IUT produce 10 key

Reviewer 2: We may wish to expand the acronym `IUT` here.

  pairs. The private key provided is used to compute the public key, Q’. The calculated value Q’ is then compared to the supplied public key, Q.

> * Public Key Validation: Generate 12 key pairs for each supported curve, modify some public keys to introduce errors, and determine whether it can detect them.

Reviewer 2: ```suggestion
* Public Key Validation: Generate 12 key pairs for each supported curve, modify some public keys to introduce errors, and verify that we can detect those errors.
```

* Signature Generation: Ten preconfigured known messages per curve supported are supplied. Generate a signature for each message. The message, public key, and signature components are returned.
  The signatures are compared against previous known signatures.

> * Signature Verification: For each curve supported, 15 pseudorandom messages, a public key, Q, and a signature component (r, s) are supplied if some of the provided values are modified so that signature verification should fail.

Reviewer 2: ```suggestion
* Signature Verification: For each curve supported, 15 pseudorandom messages, a public key, Q, and a signature component (r, s) are supplied. Modify some of the provided values and verify that signature verification fails.
```

Some ideas:
* [Cryptographic-Algorithm-Validation-Program](https://csrc.nist.gov/CSRC/media//Projects/Cryptographic-Algorithm-Validation-Program/documents/dss/ECDSAVS.pdf):
* https://csrc.nist.gov/CSRC/media/Events/lightweight-cryptography-workshop-2019/documents/papers/systematic-testing-of-lightweight-crypto-lwc2019.pdf

The proposal is to move forward with the implementation while creating a work group to discuss and collect ideas for testing and validating these features.

### Unit Tests
One of the components that can be unit-tested is native support. SPI loading and failing. JNI wrapping.
Limitations: native-support-library tests must be executed in different environments to provide comprehensive code validation.

### Performance Testing
JMH benchmarks should be provided for signature generation and aggregation.
TssMessage validation, and public key aggregation.

We should know that variance (min, max, average, median) time it takes to do the following:

Genesis for a 4, 10, 20, 40, and 100 participants.
Rekeying for a 4, 10, 20, 40, and 100 participants.
Shares Amounts distributed randomly: 200, 500, 1000, 2000, 5000

> we need to test each configuration 100 times to get confident mins, maxes, and a good average and median

Reviewer 2: We should also, per the overview, test signature _verification_ via EVM code to ensure the gas cost is rational and within requirements.

Author: As far as I understand, we haven't defined the parameters for what it means to have a rational gas cost, but we have chosen the alt-bn128 curve so that is a given (https://eips.ethereum.org/EIPS/eip-197).

But I agree I would test that the signature produced with our library is still compatible with the requirements of the EVM to be validated.

| Designers          | Austin, Cody, Edward, Rohit, Maxi, <br/> Platform team | 
| Functional Impacts | Platform team. Release engineering. DevOps             |
| Related Proposals  | TSS-Roster, TSS-Ledger-Id                              |

## Purpose and Context
A threshold signature scheme (TSS) aims to enable a threshold number of participants (shareholders) to securely and efficiently generate succinct aggregate signatures.
In our TSS implementation, a static public key should be produced that doesn't change even when the number of participants in the scheme varies,
and it can be used to verify the aggregate of a set of partial signatures produced by a threshold number of private shares.

This is important for producing proofs that are easily consumable and verifiable by external entities.

This proposal covers the implementation of all necessary components to provide the consensus node and future library users with
the functionality to sign and verify blocks using a Threshold Signature Scheme (TSS) and EC Cryptography.

The related proposal, TSS-Ledger-Id, provides an overview of the process and background for TSS and how it impacts the platform’s functionality.

> This proposal assumes no relation with the platform and defines a generic components that any consumer can integrate.

Reviewer 5: Immaterial change:
```suggestion
This proposal assumes no relation with the platform and defines a generic component that any consumer can integrate.
```

* statement: The information to prove (sidᵢs,public_keys, polynomial_commitment, encrypted_shares )

* rng: source of randomness

1. Serialize statement
2. Initialize a pseudorandom number generator using the `SHA256` hash of (1)
3. `x` = Random out of the random generator in (2)  
4. Generate random values `α`, `ρ` from the `SignatureScheme.field` out of rng
5. `g` = `SignatureScheme.publicKeyGroup.generator`
  5.1    `F = g^ρ`
  5.2    `A = g^α`
6. Calculate `Y` by transforming elements of the statement (public keys, sids) multiplied by `ρ`, then adding `A`
    
7. Serialize (`x`, `F`, `A`, `Y`)
8. Recompute the pseudorandom number generator seed using the hash of (7)
9. `x'` = Random out of the random generator in (8)
    
10. `z_r` = `x'` * `witness.randomness` + `ρ`
11. `z_a` = `x'` + `α` + Sum(`sᵢ` * `x`.pow(`sidᵢ`))
    
Return the proof composed of `F`, `A`, `Y`, `z_r`, and `z_a`

##### Outside of scope
Using an established channel, each participant will broadcast a single message to be received by all participants
while waiting to receive other participants' messages. This functionality is critical for the protocol to work but needs to be handled outside the library.

> Each participant will validate the received message against the commitment and the NIZKs proof. Invalid messages needs to be discarded.

Reviewer 5: Immaterial change:
```suggestion
Each participant will validate the received message against the commitment and the NIZKs proof. Invalid messages need to be discarded.
```

#### Security Considerations
As long as an adversary Participant knows fewer than a threshold number of decryption keys, they cannot recover enough information to start forging threshold signatures.
Adversarial participants may learn the shares of the participants whose keys they have compromised, but more is needed to recover the secret.
For security, adversarial parties might choose low or zero entropy values for protocol inputs such as shareIds, but here will still be sufficient entropy in the overall protocol from honest nodes.

#### Summary diagram
The following diagram exposes the steps necessary for bootstrapping and using the library. In orange are all the steps where the library is involved.
The rest of the steps need to happen but are outside the scope of the described library.

![img_2.png](img_2.svg)

### Architecture
To implement the functionality detailed in the previous section, the following code structure is proposed:

![img_5.png](img_5.svg)

1. **TSS Lib**: The consensus node will use the TSS library to create shares, create TSS messages to send to other nodes,

>    assemble shared public keys (ledgerId), and sign the block node Merkle tree hash.

Reviewer 5: Immaterial change - i think its best to just say that the consensus node will sign messages. We don't need to specify which.
```suggestion
   assemble shared public keys (ledgerId), and sign messages.
```

6. **Native Support Lib**: Provides a set of generic functions loading native libraries in different system architectures when packaged in a jar
   using a predefined organization so they can be accessed with JNI.
6. **[Arkworks](https://github.com/arkworks-rs)**: A Rust ecosystem for cryptography. Our implementation uses it as the library responsible for elliptic curve cryptography.
7. **EC-Key Utils** is a utility module that enables the node operator to generate a bootstrap public/private key pair.

### Module organization and repositories
![img_6.png](img_6.svg)
1. **hedera-cryptography**: This is a separate repository for hosting cryptography-related libraries.
   It is necessary to facilitate our build process, which includes Rust libraries. It also provides independent release cycles between consensus node code and other library users.
2. **swirlds-native-support**: Gradle module that enables loading into memory compiled native libraries so they can be used with JNI.
3. **swirlds-crypto-tss**: Gradle module for the TSS Library.
   This library's only client is the consensus node, so it will be in the hedera-services repository, under the `platform-sdk` folder.

> 4. **swirlds-crypto-signatures**: Gradle module for the Bilinear Pairings Signature Library.

Reviewer 5: Is this the module that is a dependency for anyone using the ledger id to verify aggregated signatures? I.e. the block node when verifying that the network has agreed on and signed the block hash?

Author: Indeed.

**Description**: A public key generated using the pairings API
###### `PairingSignature`
**Link**: [PairingSignature.java](signature-lib%2FPairingSignature.java)

**Description**: A signature generated with the private key that can be verified with the public key

###### `GroupAssignment`
**Link**: [GroupAssignment.java](signature-lib%2FGroupAssignment.java)

**Description**: An enum to clarify which group public keys and signatures are in, for a given SignatureSchema

##### Implementation Note
The serialization of the elements in this module adds a byte to represent the combination of Curve type and group assignment.

##### Examples
###### Generating a SignatureSchema
```java
SignatureSchema signatureSchema = SignatureSchema.from(Curve.ALT_BN128, GroupAssignament.G1_PUBLIC_KEYS);
```
###### Generating a Private Key
```java
   PairingPrivateKey privateKey = PairingPrivateKey.create(signatureSchema, new SecureRandom());
   
```
###### Generating a Public Key
```java
   PairingPublicKey publicKey = PairingPublicKey.create(privateKey);
   
```
###### Generating a Signature
```java
    byte[] message = new byte[]{};

>     PairingSignature signature = PairingSignature.sign(privateKey, message );

Reviewer 5: I believe the modification below is correct, since it matches the java file provided. I consider this change immaterial since the actual java API provided is correct.
```suggestion
    PairingSignature signature = PairingPrivateKey.sign(message );
```

   
```
###### Verifying a Signature
```java
static{
    signature.verify(publicKey, message);
}
```
##### Constraints
This module will not depend on hedera-services artifacts, so it cannot include logging, metrics, configuration, or any other helper module from that repo.
##### Dependencies
swirlds-cryptography-pairings-API and runtime implementation

#### Swirlds Crypto TSS Library
##### Overview
This library implements the Groth21 TSS-specific primitives.

##### Constraints
Once built, this library will depend on artifacts constructed in other repositories.
This library accepts Integer.MAX_VALUE -1 participants.

##### Public API:
###### `TssService`

> **Description**: This class handles all tss specific operations.

Reviewer 5: Will this service be stateless?

Author: Yes. No changing data structure is part of the initialization process either. We can hold a reference to the service or instantiate a new one each time it is needed.

           TssParticipantDirectory.createBuilder()
           .self(/*Identification, different for each participant*/ 0, persistentParticipantKey)
           .withParticipant(/*Identification:*/0, /*Number of Shares*/5, persistentPublicKeys.get(0))
           .withParticipant(/*Identification:*/1, /*Number of Shares*/2, persistentPublicKeys.get(1))
           .withParticipant(/*Identification:*/2, /*Number of Shares*/1, persistentPublicKeys.get(2))
           .withParticipant(/*Identification:*/3, /*Number of Shares*/1, persistentPublicKeys.get(3))
           .withThreshold(5)     
          .build(signatureScheme);
   
   //One can then query the directory
   int n = participantDirectory.getTotalNumberOfShares();
   List<TssShareId> privateShares = directory.getOwnedSharesIds();
   List<TssShareId> shareIds = directory.getShareIds();
```
###### 1. Create TssMessage
```java
   //Creates a TssMessage out of a randomly generated share 
   TssMessage message = service.generateTssMessage(directory);
```

###### 2. Validation of TssMessage
```java
static {
    List<TssMessage> messages = Requirements.receiveTssMessages();
    for(TssMessage m : messages){

>         service.validate(participantDirectory, m);

Reviewer 5: Immaterial change:
```suggestion
        service.verify(participantDirectory, m);
```

    }
}
```
###### 3. Processing of TssMessage
```java
    Set<TssMessage> agreedValidMessages = /*Some previously agreed upon same set of valid messages for all participants*/
    //Get Private Shares    
    List<TssPrivateShare> privateShares = service.decryptPrivateShares(participantDirectory, agreedValidMessages);
    //Get Public Shares

>     List<TssPrivateShare> publicShares = service.computePublicShare(participantDirectory, agreedValidMessages);

Reviewer 5: Immaterial Change:
```suggestion
    List<TssPublicShare> publicShares = service.computePublicShare(participantDirectory, agreedValidMessages);
```

>     List<TssPrivateShare> publicShares = service.computePublicShare(participantDirectory, agreedValidMessages);

Reviewer 5: Immaterial Change:
```suggestion
    TssPrivateShare publicShares = service.computePublicShare(participantDirectory, agreedValidMessages);
```

```
######  Sign
```java
   byte[] message = Requirements.messageToSign();
   List<TssShareSignature> signatures = service.sign(privateShares, message);

```

######  Aggregate Signatures

```java
static {
    List<TssShareSignature> signatures = Requirements.loadSignatures();

    //Validation of individual signatures
    List<TssShareSignature> validSignatures = new ArrayList<>(signatures.size());
    for (TssShareSignature signature : signatures) {
        if (service.verify(participantDirectory, publicShares, signature)) {
            validSignatures.add(signature);
        }
    }

    //Producing an aggregate signature

>     PairingSignature aggregateSignature = service.aggregate(participantDirectory, validSignatures);

Reviewer 5: Immaterial change:

I'm not sure if this is the public or private aggregation. Just pointing out it doesn't match the java API

- Using JCA through Bouncy Castle implementation was analyzed. BC provides support for EC-Curves, but it does not support Pairings which are required for BLS signatures.
  [beacon-chain](https://github.com/harmony-dev/beacon-chain-java/tree/master) implemented the support using bouncy castle + [Milagro](https://incubator.apache.org/projects/milagro.html) 
but Milagro project is reported to have little coverage, not audited. Milagro podling has been retired on 2024-03-21.

## Implementation and Delivery Plan by stages
**Stage 1**
* Preconditions:
  * hedera-cryptography repository.
  * Gradle multilanguage module with rust compilation plugin.
  * CI/CD pipelines for building artifacts in hedera-cryptography.
* Define the Test plan.
* Define a security plan.
* Implementation of native-support library.
* Implementation of Pairings API using JNI, arkworks, and alt-bn128.
* Execute Test Plan and validation.
* Execute Security Audits.
* Implementation of EC-key utility.

**Stage 2**
* Preconditions:
  * CI/CD pipelines to reference built artifacts in hedera-cryptography from hedera-services.

>   * Implementation of the public interface for the TSS library.

Reviewer 5: I assume this means just checking in the API that the consensus node will use. Could this be done in stage 1 so it can be used by the team working on TSS Block Signing? I do not consider this a material change.

Author: Yes, it makes sense. I was optimizing stage 1 to deliver the capacity to create the signatures for DevOps. 
But that is not the only front.
Also I realize I removed the implementation of the **swirlds-crypto-signatures** from the implementation plan by mistake.

I will create a follow-up PR for this proposal with the updated implementation plan.

Author: Sorry, misread `I do not consider this a material change`
Won't create a follow-up PR based on that.

# TSS-Library

## Summary

> Provide necessary pieces for signing messages using a TSS scheme.

Author: It is not only about a public API; it will provide at least 2 public APIs, multiple modules, implementations of those APIs, utility libraries in multiple languages, project-building logic, artifacts, and documentation.
I can use the word artifacts if that better describes the produced elements...

| Metadata           | Entities                                               | 
|--------------------|--------------------------------------------------------|
| Designers          | Austin, Cody, Edward, Rohit, Maxi, <br/> Platform team | 
| Functional Impacts | Platform team. Release engineering. DevOps             |
| Related Proposals  | TSS-Roster, TSS-Ledger-Id                              |
| HIPS               | ?                                                      |

## Purpose and Context

> A threshold signature scheme (TSS) aims to enable a threshold number of participants (shareholders) to securely and efficiently generate succinct aggregate signatures.

Reviewer 6: succinct?

Author: short

In our TSS implementation, a static public key should be produced that doesn't change even when the number of participants in the scheme varies,
and it can be used to verify the aggregate of a set of partial signatures produced by a threshold number of private shares.

This is important for producing proofs that are easily consumable and verifiable by external entities.

> This proposal covers the implementation of all necessary pieces to provide the consensus node and block node with

Author: changed for components

2. **swirlds-native-support**: Gradle module that enables loading into memory compiled native libraries so they can be used with JNI.
3. **swirlds-cryptography-tss**: Gradle module for the TSS Library.
   This library's only client is the consensus node, so it will be in the hedera-services repository, under the `platform-sdk` folder.
4. **swirlds-cryptography-signatures**: Gradle module for the Bilinear Pairings Signature Library.
5. **swirlds-cryptography-pairings-api**: Gradle module for the Bilinear Pairings API. Minimizes the impact of adding or removing implementations.
6. **swirlds-cryptography-alt128**: Gradle module that will implement the Bilinear Pairings API using alt-128 elliptic curve.
   That curve has been chosen due to EVM support. The artworks rust library will provide the underlying cryptography implementation.
   The module will include Java and Rust codes that will be compiled for all possible system architectures and distributed in a jar under a predefined structure.

### Libraries Specifications

#### Swirlds Crypto TSS Library
##### Overview
This library implements the Groth21 TSS-specific primitives.

> Two options are presented, while the first focus on usability from the client perspective, the second one proposes operations that are more atomic focusing on providing clients with the ability of concurrently handle the library's operation.

Reviewer 7: > while the first focus on usability from the client perspective

Nothing wrong with describing different options, but IMO option 1 is _not_ more usable from the client's perspective. The client is the platform, and these shared data structures make the library more difficult to use by the platform.

Author: @alittley ended up merging both proposals into one.
As I said, there are no mutable shared data structures.

Reviewer 7: This seems reasonable.

###### `TssMessage`
**Description**: This class is used in the exchange of secret information between participants. Contains an encrypted message, a polynomial commitment, and a cryptographic proof that can be used to validate this message.

**Link**:[TssMessage.java](tss-option1%2FTssMessage.java)

###### `TssShareId`
**Description**: This class represents the unique identification of a share in the system. It a FieldElement, so it can be used as input in the polynomial.

**Link**:[TssShareId.java](tss-option1%2FTssShareId.java)

###### `TssPrivateShare`
**Description**: A record that contains a share ID, and the corresponding private key.

**Link**:[TssPrivateShare.java](tss-option1%2FTssPrivateShare.java)

###### `TssPublicShare`
**Description**: A record that contains a share ID, and the corresponding public key.

**Link**:[TssPublicShare.java](tss-option1%2FTssPublicShare.java)

###### `TssShareSignature`
**Description**: Represents a signature created for a TSSPrivateShare.

**Link**:[TssShareSignature.java](tss-option1%2FTssShareSignature.java)

##### Option 1 - Examples:
###### 1. Bootstrapping
```java
   TssService service = new TssService(signatureScheme, new Random());

>    PairingPrivateKey persistentParticipantKey = Requirements.loadECPrivateKey();

Author: It is not defined; it just signifies that information is coming from the outside. It is required information. 
I aimed to produce code that compiled in the examples. 
Could be:
```
 PairingPrivateKey persistentParticipantKey = /*SOME KEY*/
```

   List<PairingPublicKey> persistentPublicKeys = Requirements.loadECPublicKeys();
   TssParticipantDirectory participantDirectory =
           TssParticipantDirectory.createBuilder()
           .self(/**/ 0, persistentParticipantKey)
           .withParticipant(/*Identification:*/0, /*Number of Shares*/5, persistentPublicKeys.get(0))
           .withParticipant(/*Identification:*/1, /*Number of Shares*/2, persistentPublicKeys.get(1))
           .withParticipant(/*Identification:*/2, /*Number of Shares*/1, persistentPublicKeys.get(2))
           .withParticipant(/*Identification:*/3, /*Number of Shares*/1, persistentPublicKeys.get(3))
           .withThreshold(5)     
          .build(signatureScheme);
   
   //One can then query the directory

>    int n = directory.getTotalNumberOfShares();

Author: it is! thanks!

##### Option 2 - Public API:
###### `ShareClaims`

**Link**: [ShareClaims.java](tss-option2%2FShareClaims.java)

###### `Tss`

**Link**:[Tss.java](tss-option2%2FTss.java)

###### `TssMessage`

**Link**: [TssMessage.java](tss-option2%2FTssMessage.java)

###### `TssPrivateShare`

**Link**:[TssPrivateShare.java](tss-option2%2FTssPrivateShare.java)

###### `TssPublicShare`

**Link**:[TssPublicShare.java](tss-option2%2FTssPublicShare.java)

###### `TssShareClaim`
**Link**: [TssShareClaim.java](tss-option2%2FTssShareClaim.java)

###### `TssShareId`
**Link**: [TssShareId.java](tss-option2%2FTssShareId.java)

###### `TssShareSignature`
**Link**: [TssShareSignature.java](tss-option2%2FTssShareSignature.java)

##### Option 2 - Examples:
###### Get a concrete Tss Implementation out of a [SignatureSchema](#Swirlds-Cryptography-Pairings-Signature-Library)
```java
  Tss tssImpl = TSS.getFor(signatureSchema);
```

######  Creating TssShareClaims

```java
PairingPublicKey p1PbK = Requirements.loadECPublicKey("P1");
PairingPublicKey p2PbK = Requirements.loadECPublicKey("P2");

int p1Shares = Requirements.numberOfShares("P1");
int p2Shares = Requirements.numberOfShares("P2");

> List<TssShareClaim> claims = new ArrayList<>(p1Shares + p2Shares);

Author: That is the initial capacity of the list, only to showcase that the elements inside the list will be 1 per total share.
The elements added to the list are 1 per p1Shares, and 1 per  p2Shares. The initial capacity of the list is p1Shares + p2Shares

> List<TssShareClaim> claims = new ArrayList<>(p1Shares + p2Shares);

Reviewer 8: I know that but in the following 2 loops you add more than (`p1Shares+p1Shares`) elements to the list.
Assume `p1Shares=2` and `p2Shares=5`. By doing so you would define the inital size as 7.

The first loop will add `p1Shares` (2) elements. The second loop will add `p1Shares+p1Shares` (7) elements. This ends in 9 elements.

> List<TssShareClaim> claims = new ArrayList<>(p1Shares + p2Shares);

Reviewer 8: sure. but the 2 following loops will add more elements to the list than you defined with the initial capacity

```
######  Signing messages
```java
//for each private share of the node:
for (TssPrivateShare share : privateShares){
    TssSignature signature = share.sign(blockHash);
    Platform.send(signature);
}

//collect signatures
List<TssShareSignature> collectedSignatures = Requirements.collectTssSignatures();
List<TssShareSignature> validSignatures = new ArrayList<>();
for (TssShareSignature signature : collectedSignatures){
    if(signature.signature().verifySignature(tssShareIdTssPublicShareMap.get(signature.shareId()).publicKey(),
blockHash)){
        validSignatures.add(signature);
            }
}
```
######  Aggregate Signatures
```java
//collect signatures
List<TssShareSignature> collectedSignatures = Requirements.collectTssSignatures();
List<TssShareSignature> validSignatures = new ArrayList<>();
for (TssShareSignature signature : collectedSignatures){
    if(signature.signature().verifySignature(tssShareIdTssPublicShareMap.get(signature.shareId()).publicKey(),
blockHash)){
        validSignatures.add(signature);
            }
}

PairingSignature value = Tss.aggregateSignatures(validSignatures);
```

#### Swirlds Crypto Pairings Signature Library
##### Overview

> This module provides cryptography primitives to create EC PublicKeys, EC PrivateKeys, and Signatures.

Author: It is true.
There is the bottom-up approach and the top-down approach.
Originally I followed the bottom-up approach, pairings, signature, tss.  So each module relied on pieces already explained. But then, when I added the general explanation of the process in the TSS library, it seemed wrong that it was the last topic in the document. So I went with the top-down approach.

> This module provides cryptography primitives to create EC PublicKeys, EC PrivateKeys, and Signatures.

Reviewer 8: In that case it would make sense to mention that the interfaces will be described later.

##### Public API
###### `SignatureSchema`
**Description**: A pairings signature scheme can be implemented with different types of curves and group assignment configurations.
For example, two different configurations might consist of a BLS_12_381 curve using G1 of the pairing to generate public key elements or G2 for the same purpose.

> ###### `PairingPrivateKey`

Author: That was on purpose; they are simply records whose structure can be inferred by the example code.
I can add them to the code folder, but I deemed that doing so would not add substantial information to the proposal.

> ###### `PairingPrivateKey`

Author: Added them

**Description**: A private key generated using the pairings API
###### `PairingPublicKey`
**Description**: A public key generated using the pairings API
###### `PairingSignature`
**Description**: A signature generated with the private key that can be verified with the public key

##### Implementation Note
The serialization of the elements in this module adds a byte to represent the combination of Curve type and group assignment.

##### Examples
###### Generating a SignatureSchema
```java

> SignatureSchema signatureSchema = SignatureSchema.from(Curve.ALT_BN128, GroupAssignament.G1_PUBLIC_KEYS);

Author: Thanks forgot to include it.

**Link**: [FieldElement.java](pairings-api%2FFieldElement.java)

###### `Group`

**Description**: This class provides methods to obtain elements belonging to the group represented by the instance.

**Link**: [Group.java](pairings-api%2FGroup.java)

###### `GroupElement`

**Description**: An interface representing an element within a group, providing methods for basic group operations.

**Link**: [GroupElement.java](pairings-api%2FGroupElement.java)

###### `PairingResult`

**Description**: An interface representing the result of a pairing operation, with methods to compare it to other group elements.

**Link**: [PairingResult.java](pairings-api%2FPairingResult.java)

#### Swirlds Crypto Pairings Impl
##### Overview
This module will be a multilanguage module following the structure:
```
 swirlds-crypto-pairings-impl-X
 |_src
    |_main
    | |_java
    | | |_com...
    | |_rust
    |   |_...
    |_test
      |_java
      | |_com...
      |_rust
        |_...
```
Support for release engineering will be needed.
This module should include Cargo for compiling rust source to binary files,

> It should be built in a way that the rust code is compiled for every supported architecture and only after that, packaged into the jar following the structure:

Author: This was decided in a meeting on June 4th with @nathanklick and @rbarkerSL.

More in detail, the reasoning behind this decision is to combine all things with the same reason to change. 
Given that we are using JNI, it requires code in both languages (java and Rust) that will change together. If we separate that unit of change into different modules, we will produce difficult workflows for developers.

Indeed, a Windows machine cannot be used to build for other architectures unless docker is involved. The proposal also mentions that on a developer machine, the produced jar will have only the version for that machine. This is a significant change considering our built logic today.
However, the artifact built and published to Maven should support all libraries, so it is transparent that dev is only using the libraries and not modifying them.

> It should be built in a way that the rust code is compiled for every supported architecture and only after that, packaged into the jar following the structure:

Reviewer 8: I still don't understand how you will build an artefact containing native compilations for n architectures and operation systems in justst, that should be one step on one machine. At least the solution should be described here in a general way. With that definition I would currently vote -1 since I have no clue how that can be done without way too much complexity.

> It should be built in a way that the rust code is compiled for every supported architecture and only after that, packaged into the jar following the structure:

Author: After discussing, we agreed that things beyond the proposal's scope should not be detailed at a deep level.
I re-clarified that other teams will provide that part, so the document is clearer about that point.

```
 |_WEB-INF
    |_arch64
      |_macos
        |_lib.dylib
    |_amd64
      |_macos
        |_lib.dylib
      |_linux
        |_lib.so
      |_windows
        |_lib.dll
    |_x86
      |_linux
        |_lib.so
      |_windows
        |_lib.dll
```
It will unzip and load the correct library packaged inside the jar and access it using JNI calls.
The API interface is through JNI. Once built, this should become a runtime dependency.

#### Swirlds Native Support
##### Overview
Our implementation of the `swirlds-cryptography-pairings-api` will use native compiled libraries under the hood accessed with JNI.

> Given the non-GA state of Project Panama, this proposal does not consider it, but it has not been discarded for future change proposals.

Author: As far as I know, we cannot wait for Panama to be GA to implement this. But you can take that to @jasperpotts and @poulok.
We have POC [task for analyzing Panama](https://github.com/hashgraph/hedera-services/issues/14014) that can be taken independently of this proposal. Still, I would not delay implementing what we can today with the POC that analyzes how to implement this in the future.

> Given the non-GA state of Project Panama, this proposal does not consider it, but it has not been discarded for future change proposals.

Reviewer: The next LTS to include the new FFI capability is September of 2025.  https://www.oracle.com/java/technologies/java-se-support-roadmap.html

That is too long to wait.

> Given the non-GA state of Project Panama, this proposal does not consider it, but it has not been discarded for future change proposals.

Reviewer 8: I'm not saying that we should wait. We should define an API / workflow today that can easy be refactored to FFI once it is availble. We should not ignore it but think about it when defining everything based on JNI

   and reference its location in the classpath env variable.
2. Distribute it as an application dependency. The caveat here is that:

   a. The distributed library must be compiled in every possible architecture on which the application will be executed.

   b. The library must be unpackaged from the jar as it cannot be accessed while compressed.

Two key points we are considering in the development of this solution are:
a) ensuring that it is compatible with the Java module system.
b) Ensure that dependent software does not require the installation of any additional dependencies other than what is distributed in the jars.

That is why the native-support library will help load the native library by implementing option 2.
For this to work, the native support library must be included as a dependency in the module accessing the JNI wrapping code.
The native libraries will be accessed assuming the following structure in the module declaring the dependency:

```
 native-library-client.jar
 |_WEB-INF
    |_arch64
      |_macos
        |_libhedera_bls_jni.dylib
    |_amd64
      |_macos

>         |_libhedera_bls_jni.dylib

Reviewer 8: just an idea

      |_linux
        |_libhedera_bls_jni.so
      |_windows
        |_libhedera_bls_jni.dll
    |_x86
      |_linux
        |_libhedera_bls_jni.so
      |_windows
        |_libhedera_bls_jni.dll
  ...
```

lib folder is organized in subfolders by the platform identifier, as returned by `System.getProperty("os.arch")` and `System.getProperty("os.name")`

##### Constraints
This module will not depend on hedera-services artifacts, so it cannot include logging, metrics, configuration, or any other helper module from that repo.

##### Public API
###### `LibraryDescriptionEntry`
**Description**: Given that the compilation of a native library produces files with different names under different OS and architectures, we need a way to assemble a catalog for all possible forms our library will take.

A record of 3 elements that defines the name of the binary file of the library to load in a specific system architecture and OS.

**Example**:
```java

>     new LibraryDescriptionEntry("macos", "aarch64", "libhedera_bls_jni.dylib");

Reviewer 8: We should have constants for common OS and ARCH as part of the public API. Next to that `String getOs()` / `String getArch()` would be nive as wrappers around System.getProperty.

Author: I agree. Adding that.

```
###### `LibraryDescription`
**Description**: A description of the library in all possible systems.
```java
    static final LibraryDescription LIB_HEDERA_BLS = new LibraryDescription(new LibraryDescriptionEntry("macos", "aarch64", "libhedera_bls_jni.dylib"), new LibraryDescriptionEntry("linux", "amd64", "libhedera_bls_jni.so"), ...);
```
If the library name is the same for all system architectures with the only change of the extension, one can configure:
```java
  //For any system architecture load libhedera_bls_jni.dylib, libhedera_bls_jni.so or libhedera_bls_jni.dll depending on the os  
  static final LibraryDescription LIB_HEDERA_BLS = new LibraryDescription("libhedera_bls_jni");
```

###### `LibraryLoader`

> **Description**: Helper class that will load a library for the correct system:

Author: I'll add that

**Description**: This is a factory interface responsible for creating FieldElement, which are scalars belonging to the field represented by this instance.

**Link**: [Field.java](pairings-api%2FField.java)

###### `FieldElement`

**Description**: An interface representing an element within a field, providing methods for basic arithmetic operations.

**Link**: [FieldElement.java](pairings-api%2FFieldElement.java)

###### `Group`

**Description**:  An interface to define methods that obtain elements belonging to the group(s) supported by an implementation.

**Link**: [Group.java](pairings-api%2FGroup.java)

###### `GroupElement`

**Description**: An interface representing an element within a group, providing methods for basic group operations.

**Link**: [GroupElement.java](pairings-api%2FGroupElement.java)

###### `PairingResult`

**Description**: An interface representing the result of a pairing operation, with methods to compare it to other group elements.

**Link**: [PairingResult.java](pairings-api%2FPairingResult.java)

###### `PairingResult`

**Description**: An interface representing the result of a pairing operation, with methods to compare it to other group elements.

> **Link**: [PairingResult.java](pairings-api%2FPairingResult.java)

Reviewer 2: Duplicate section

- **Share**: Represents a piece of the necessary public/private elements to create signatures. In TSS,
  a threshold number of shares is needed to produce an aggregate signature that the ledger public key can later verify.
- **Polynomial Commitment**: It enables verification of evaluations of the polynomial at specific points without revealing the entire polynomial.

> - **Participant**: Any party involved in the distributed key generation protocol. Their goal is to generate a public key and provide a set of receivers with matching secret shares of the secret key. 

Reviewer 7: > Their goal is to generate a public key and provide a set of receivers with matching secret shares of the secret key.

This description is confusing to me. Is the public key you are referring to the ledger ID? What is "matching" about the secret shares?

IMO it's not necessary to state a goal of the participant here, since the goal is just to do the protocol described in this doc.

Author: I removed the goal part. 
In my head, the explanation made more sense than what ended up in the paper. I was thinking that the secret should be split into matching pieces for each share and encrypted with the matching participant's public key.

##### 1. Create TssMessage
`TssMessage`: A data structure for distributing encrypted shares of a secret among all participants in a way that only the intended participant can see its part of the share.

> It includes auxiliary information used to validate its correctness and assemble an aggregate public key ie: a commitment to a secret share polynomial and a NIZK proof.

Reviewer 7: "public key ie: a"... -> "public key, i.e., a"...

**Description**: This class is used in the exchange of secret information between participants. Contains an encrypted message, a polynomial commitment, and a cryptographic proof that can be used to validate this message.
###### `TssShareId`
**Description**: This class represents the unique identification of a share in the system. It a FieldElement, so it can be used as input in the polynomial.
###### `TssPrivateShare`
**Description**: A record that contains a share ID, and the corresponding private key.
###### `TssPublicShare`
**Description**: A record that contains a share ID, and the corresponding public key.
###### `TssShareSignature`

> **Description**: Represents a signature created for a TSSPrivateShare.

Reviewer 7: created _from_ a TssPrivateShare

After genesis or rekeying stages, the library can sign any message.
Using each `SecretShare` owned by the participant, a message can be signed, producing a `TssSignature`

Multiple signatures can be aggregated to create an aggregate `TssSignature`. An aggregate `TssSignature` can be validated against the LedgerId (public key)  if
`t` (t=threshold) valid signatures are aggregated. If the threshold is met and the signature is valid, the library will respond with true; if not, it will respond with false.

#### Security Considerations
As long as an adversary Participant knows fewer than a threshold number of decryption keys, they cannot recover enough information to start forging threshold signatures.
Adversarial participants may learn the shares of the participants whose keys they have compromised, but more is needed to recover the secret.

> Suggested change

Reviewer 2: Accidental copy/paste error?

#### Security Considerations
As long as an adversary Participant knows fewer than a threshold number of decryption keys, they cannot recover enough information to start forging threshold signatures.
Adversarial participants may learn the shares of the participants whose keys they have compromised, but more is needed to recover the secret.
For security, adversarial parties are free to choose low or zero entropy values for protocol inputs such as shareIds.
There will still be sufficient entropy in the entire protocol from the honest nodes.

> For security, adversarial parties might choose low or zero entropy values for protocol inputs such as shareIds, but here will still be sufficient entropy in the overall protocol from honest nodes.

Reviewer 2: Still duplicating lines; I think the first two are the old version, the last line is the revision.

###### `BilinearPairing`

**Description**: This class provides access to each group (G₁, G₂) for a specific Pairing and the FiniteField associated with the curve.

> It provides an itialization method to load all necessary dependencies for the library to work. Callers are requested to invoke that method before start using the library.

Reviewer 2: `itialization` ==> `initialization`

###### `PairingResult`

**Description**: An interface representing the result of a pairing operation, with methods to compare it to other group elements.

**Link**: [PairingResult.java](pairings-api%2FPairingResult.java)

##### Examples:
###### Get a Pairing instance for a given curve
```java
static {
    //This will get an initialized instance of the pairing 
    BilinearPairing pairing = BilinearPairingService.instanceOf(Curve.ALT_BN128);
}
```
###### Work with the field associated to the pairing

```java

static {
    final Field field = pairing.getField();
    final FieldElement indexElement = field.elementFromLong(10L);
    final FieldElement randomElement = field.randomElement(new SecureRandom());
    final FieldElement otherElement = indexElement.power(BigInteger.valueOf(100));
    final byte[] seed = new byte[getSeedSize()];
    random.nextBytes(seed);
    final FieldElement anotherElement = field.randomElement(seed);
    final FieldElement resultAddition = randomElement.add(anotherElement);
    final FieldElement resultMultiplication = randomElement.multiply(anotherElement);
}
```
###### Work with the groups associated to the pairing
```java

> import javax.swing.GroupLayout.Group;

Reviewer 2: You probably did not mean to bring in the Swing component for `Group`.
